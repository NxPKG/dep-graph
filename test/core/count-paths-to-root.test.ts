import * as depGraphLib from '../../src';
import * as helpers from '../helpers';

describe('countPathsToRoot', () => {
  describe('basic', () => {
    const depGraphData = helpers.loadFixture('plain-dep-graph.json');
    const depGraph = depGraphLib.createFromJSON(depGraphData);

    it('returns expected path counts for all packages', () => {
      const counts: Record<string, number> = {};
      for (const pkg of depGraph.getPkgs()) {
        counts[`${pkg.name}@${pkg.version}`] = depGraph.countPathsToRoot(pkg);
      }
      expect(counts).toEqual({
        'root@0.0.0': 1,
        'e@5.0.0': 7,
        'd@0.0.1': 1,
        'c@1.0.0': 2,
        'a@1.0.0': 1,
        'd@0.0.2': 2,
        'b@1.0.0': 1,
        'i@2.1.0': 4,
        'h@0.0.1': 3,
        'g@1.0.0': 2,
        'f@1.0.0': 1,
      });
    });

    it('returns limited path counts for all packages', () => {
      const counts: Record<string, number> = {};
      for (const pkg of depGraph.getPkgs()) {
        counts[`${pkg.name}@${pkg.version}`] = depGraph.countPathsToRoot(pkg, {
          limit: 2,
        });
      }
      expect(counts).toEqual({
        'root@0.0.0': 1,
        'e@5.0.0': 2,
        'd@0.0.1': 1,
        'c@1.0.0': 2,
        'a@1.0.0': 1,
        'd@0.0.2': 2,
        'b@1.0.0': 1,
        'i@2.1.0': 2,
        'h@0.0.1': 2,
        'g@1.0.0': 2,
        'f@1.0.0': 1,
      });
    });
  });

  describe('large', () => {
    const depGraphData = helpers.loadFixture('goof-graph.json');

    it('returns expected path counts for all packages', () => {
      const depGraph = depGraphLib.createFromJSON(depGraphData);
      const counts: Record<string, number> = {};
      for (const pkg of depGraph.getPkgs()) {
        counts[`${pkg.name}@${pkg.version}`] = depGraph.countPathsToRoot(pkg);
      }
      expect(counts).toMatchSnapshot();
    });

    it('returns limited path counts for all packages', () => {
      const depGraph = depGraphLib.createFromJSON(depGraphData);
      for (const pkg of depGraph.getPkgs()) {
        expect(
          depGraph.countPathsToRoot(pkg, { limit: 2 }),
        ).toBeLessThanOrEqual(2);
      }
    });

    it('returns identical limited paths counts with and without internal cache', () => {
      // One: run with a limit on a fresh dep-graph
      const depGraph = depGraphLib.createFromJSON(depGraphData);
      const countsWithoutCache: Record<string, number> = {};
      for (const pkg of depGraph.getPkgs()) {
        countsWithoutCache[`${pkg.name}@${pkg.version}`] =
          depGraph.countPathsToRoot(pkg, { limit: 2 });
      }

      // Two: run without a limit on a new dep-graph
      const depGraphB = depGraphLib.createFromJSON(depGraphData);
      for (const pkg of depGraphB.getPkgs()) {
        depGraphB.countPathsToRoot(pkg);
      }

      // Three: Use that same dep-graph instance to run with a limit
      const countsWithCache: Record<string, number> = {};
      for (const pkg of depGraphB.getPkgs()) {
        countsWithCache[`${pkg.name}@${pkg.version}`] =
          depGraphB.countPathsToRoot(pkg, { limit: 2 });
      }
      expect(countsWithCache).toEqual(countsWithoutCache);
    });
  });

  describe('cycles', () => {
    const depGraphData = helpers.loadFixture('cyclic-complex-dep-graph.json');
    const depGraph = depGraphLib.createFromJSON(depGraphData);

    it('returns 1 for the root node', () => {
      expect(depGraph.countPathsToRoot(depGraph.rootPkg)).toBe(1);
    });

    it.each`
      name   | version | expected
      ${'a'} | ${'1'}  | ${1}
      ${'b'} | ${'2'}  | ${1}
      ${'c'} | ${'3'}  | ${1}
      ${'d'} | ${'4'}  | ${1}
      ${'e'} | ${'5'}  | ${3}
      ${'f'} | ${'6'}  | ${3}
      ${'g'} | ${'7'}  | ${3}
    `('returns $expected for $name@$version', ({ name, version, expected }) => {
      expect(depGraph.countPathsToRoot({ name, version })).toBe(expected);
    });

    it.each(depGraph.getPkgs())(`equals pkgPathsToRoot(%s).length`, (pkg) => {
      expect(depGraph.countPathsToRoot(pkg)).toBe(
        depGraph.pkgPathsToRoot(pkg).length,
      );
    });
  });
});
