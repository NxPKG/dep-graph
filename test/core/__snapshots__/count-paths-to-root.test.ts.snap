// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countPathsToRoot large returns expected path counts for all packages 1`] = `
Object {
  "abbrev@1.0.7": 2,
  "abbrev@1.1.1": 1,
  "accepts@1.1.4": 1,
  "accepts@1.2.13": 1,
  "adm-zip@0.4.7": 1,
  "align-text@0.1.4": 2,
  "amdefine@1.0.0": 2,
  "ansi-regex@2.0.0": 4,
  "ansi-regex@2.1.1": 3,
  "ansi-styles@2.2.1": 1,
  "append-transform@0.4.0": 1,
  "argparse@1.0.10": 6,
  "argparse@1.0.7": 1,
  "arr-diff@2.0.0": 2,
  "arr-flatten@1.0.1": 2,
  "array-unique@0.2.1": 2,
  "arrify@1.0.1": 2,
  "asn1@0.1.11": 1,
  "asn1@0.2.4": 1,
  "assert-plus@0.1.5": 1,
  "assert-plus@0.2.0": 1,
  "assert-plus@1.0.0": 5,
  "async-cache@0.1.5": 1,
  "async@0.2.10": 1,
  "async@0.9.0": 2,
  "async@1.5.2": 2,
  "asynckit@0.4.0": 1,
  "aws-sign2@0.5.0": 1,
  "aws-sign2@0.6.0": 1,
  "aws4@1.8.0": 1,
  "balanced-match@0.4.1": 5,
  "balanced-match@1.0.0": 3,
  "basic-auth@2.0.1": 1,
  "bcrypt-pbkdf@1.0.2": 1,
  "bl@0.9.5": 1,
  "bluebird@2.9.26": 1,
  "bluebird@3.5.2": 2,
  "body-parser@1.9.0": 1,
  "boom@0.4.2": 2,
  "boom@2.10.1": 2,
  "brace-expansion@1.1.11": 3,
  "brace-expansion@1.1.4": 5,
  "braces@1.8.5": 2,
  "bson@0.4.23": 2,
  "builtin-modules@1.1.1": 2,
  "busboy@0.2.14": 1,
  "bytes@1.0.0": 2,
  "caching-transform@1.0.1": 1,
  "camelcase@1.2.1": 1,
  "camelcase@2.1.1": 1,
  "camelcase@3.0.0": 1,
  "caseless@0.11.0": 1,
  "caseless@0.6.0": 1,
  "center-align@0.1.3": 1,
  "cfenv@1.1.0": 1,
  "chalk@1.1.3": 1,
  "clean-yaml-object@0.1.0": 1,
  "cliui@2.1.0": 1,
  "cliui@3.2.0": 1,
  "code-point-at@1.0.0": 3,
  "codecov.io@0.1.6": 1,
  "color-support@1.1.3": 1,
  "combined-stream@0.0.7": 1,
  "combined-stream@1.0.7": 2,
  "commander@2.19.0": 1,
  "commondir@1.0.1": 1,
  "concat-map@0.0.1": 8,
  "config-chain@1.1.12": 1,
  "connect-busboy@0.0.2": 1,
  "consolidate@0.14.5": 1,
  "content-disposition@0.5.0": 1,
  "content-type@1.0.4": 1,
  "convert-source-map@1.2.0": 1,
  "cookie-parser@1.3.3": 1,
  "cookie-signature@1.0.5": 1,
  "cookie-signature@1.0.6": 1,
  "cookie@0.1.2": 2,
  "core-util-is@1.0.2": 9,
  "coveralls@2.13.3": 1,
  "crc@3.2.1": 3,
  "cross-spawn-async@2.2.4": 2,
  "cross-spawn@4.0.2": 1,
  "cryptiles@0.2.2": 1,
  "cryptiles@2.0.5": 1,
  "ctype@0.5.3": 1,
  "dashdash@1.14.1": 1,
  "debug@2.2.0": 6,
  "debug@2.6.9": 1,
  "debug@3.1.0": 1,
  "decamelize@1.2.0": 2,
  "deep-equal@0.1.2": 1,
  "deep-is@0.1.3": 1,
  "deeper@2.1.0": 1,
  "default-require-extensions@1.0.0": 2,
  "defined@0.0.0": 1,
  "delayed-stream@0.0.5": 1,
  "delayed-stream@1.0.0": 2,
  "depd@1.0.1": 4,
  "depd@1.1.2": 1,
  "destroy@1.0.3": 2,
  "dicer@0.2.5": 1,
  "diff@1.4.0": 1,
  "duplexer@0.1.1": 1,
  "dustjs-helpers@1.5.0": 1,
  "dustjs-linkedin@2.5.0": 1,
  "ecc-jsbn@0.1.2": 1,
  "ee-first@1.0.5": 1,
  "ee-first@1.1.0": 4,
  "ee-first@1.1.1": 1,
  "ejs-locals@1.0.2": 1,
  "ejs@0.8.8": 1,
  "ejs@1.0.0": 1,
  "error-ex@1.3.0": 3,
  "errorhandler@1.2.0": 1,
  "es6-promise@2.1.1": 1,
  "escape-html@1.0.1": 6,
  "escape-string-regexp@1.0.5": 2,
  "escodegen@1.8.0": 1,
  "esprima@2.7.2": 3,
  "esprima@2.7.3": 1,
  "esprima@4.0.1": 5,
  "estraverse@1.9.3": 1,
  "esutils@2.0.2": 1,
  "etag@1.6.0": 3,
  "events-to-array@1.1.2": 2,
  "expand-brackets@0.1.5": 2,
  "expand-range@1.8.2": 2,
  "express-fileupload@0.0.5": 1,
  "express@4.12.4": 1,
  "extend@3.0.2": 1,
  "extglob@0.3.2": 2,
  "extsprintf@1.3.0": 2,
  "fast-levenshtein@1.1.3": 1,
  "fd@0.0.3": 1,
  "file-type@8.1.0": 1,
  "filename-regex@2.0.0": 2,
  "fileset@0.2.1": 1,
  "fill-range@2.2.3": 2,
  "finalhandler@0.3.6": 1,
  "find-cache-dir@0.1.1": 1,
  "find-up@1.1.2": 6,
  "for-in@0.1.5": 2,
  "for-own@0.1.4": 2,
  "foreground-child@1.5.1": 2,
  "foreground-child@1.5.6": 1,
  "forever-agent@0.5.2": 1,
  "forever-agent@0.6.1": 1,
  "form-data@0.1.4": 1,
  "form-data@2.1.4": 1,
  "forwarded@0.1.2": 1,
  "fresh@0.2.4": 3,
  "fs-extra@0.22.1": 1,
  "fs.realpath@1.0.0": 3,
  "generate-function@2.3.1": 1,
  "generate-object-property@1.2.0": 1,
  "getpass@0.1.7": 1,
  "glob-base@0.3.0": 2,
  "glob-parent@2.0.0": 2,
  "glob@5.0.15": 1,
  "glob@7.0.3": 3,
  "glob@7.1.3": 3,
  "goof@0.0.3": 1,
  "graceful-fs@1.2.3": 1,
  "graceful-fs@4.1.11": 2,
  "graceful-fs@4.1.4": 6,
  "handlebars@4.0.5": 1,
  "har-validator@2.0.6": 1,
  "has-ansi@2.0.0": 1,
  "has-flag@1.0.0": 1,
  "hawk@1.1.1": 1,
  "hawk@3.1.3": 1,
  "hoek@0.9.1": 4,
  "hoek@2.16.3": 4,
  "hooks-fixed@1.1.0": 1,
  "hosted-git-info@2.1.5": 2,
  "http-signature@0.10.1": 1,
  "http-signature@1.1.1": 1,
  "humanize-ms@1.0.1": 1,
  "iconv-lite@0.4.4": 2,
  "imurmurhash@0.1.4": 1,
  "inflight@1.0.5": 4,
  "inflight@1.0.6": 3,
  "inherits@1.0.2": 1,
  "inherits@2.0.1": 4,
  "inherits@2.0.3": 14,
  "ini@1.1.0": 1,
  "ini@1.3.5": 1,
  "invert-kv@1.0.0": 1,
  "ipaddr.js@1.0.5": 1,
  "is-arrayish@0.2.1": 3,
  "is-buffer@1.1.3": 10,
  "is-builtin-module@1.0.0": 2,
  "is-dotfile@1.0.2": 2,
  "is-equal-shallow@0.1.3": 2,
  "is-extendable@0.1.1": 2,
  "is-extglob@1.0.0": 14,
  "is-fullwidth-code-point@1.0.0": 3,
  "is-glob@2.0.1": 8,
  "is-my-ip-valid@1.0.0": 1,
  "is-my-json-valid@2.19.0": 1,
  "is-number@2.1.0": 4,
  "is-posix-bracket@0.1.1": 2,
  "is-primitive@2.0.0": 4,
  "is-property@1.0.2": 2,
  "is-typedarray@1.0.0": 1,
  "is-utf8@0.2.1": 5,
  "isarray@0.0.1": 5,
  "isarray@1.0.0": 5,
  "isexe@1.1.2": 7,
  "isexe@2.0.0": 1,
  "isobject@2.1.0": 2,
  "isstream@0.1.2": 1,
  "istanbul@0.4.3": 1,
  "jquery@2.2.4": 1,
  "js-yaml@3.11.0": 5,
  "js-yaml@3.6.1": 2,
  "jsbn@0.1.1": 2,
  "json-schema@0.2.3": 1,
  "json-stringify-safe@5.0.1": 2,
  "jsonfile@2.4.0": 1,
  "jsonify@0.0.0": 1,
  "jsonpointer@4.0.1": 1,
  "jsprim@1.4.1": 1,
  "kareem@1.0.1": 1,
  "kerberos@0.0.24": 1,
  "kind-of@3.0.3": 10,
  "lazy-cache@1.0.4": 1,
  "lcid@1.0.0": 1,
  "lcov-parse@0.0.10": 1,
  "levn@0.3.0": 1,
  "load-json-file@1.1.0": 3,
  "lodash.assign@4.0.9": 3,
  "lodash.keys@4.0.7": 3,
  "lodash.rest@4.0.3": 3,
  "log-driver@1.2.5": 1,
  "longest@1.0.1": 2,
  "lru-cache@2.3.1": 1,
  "lru-cache@4.0.1": 2,
  "lru-cache@4.1.3": 1,
  "marked@0.3.5": 1,
  "md5-hex@1.3.0": 2,
  "md5-o-matic@0.1.1": 2,
  "media-typer@0.3.0": 3,
  "merge-descriptors@1.0.0": 1,
  "method-override@3.0.0": 1,
  "methods@1.1.2": 2,
  "micromatch@2.3.8": 2,
  "mime-db@1.12.0": 2,
  "mime-db@1.37.0": 4,
  "mime-types@1.0.2": 1,
  "mime-types@2.0.14": 2,
  "mime-types@2.1.21": 4,
  "mime@1.2.11": 2,
  "mime@1.3.4": 2,
  "minimatch@2.0.10": 2,
  "minimatch@3.0.0": 3,
  "minimatch@3.0.4": 3,
  "minimist@0.0.10": 1,
  "minimist@0.0.8": 5,
  "minimist@1.2.0": 1,
  "mkdirp@0.3.5": 1,
  "mkdirp@0.5.1": 5,
  "moment@2.15.1": 1,
  "mongodb-core@1.2.19": 1,
  "mongodb@2.0.46": 1,
  "mongoose@4.2.4": 1,
  "morgan@1.9.1": 1,
  "mpath@0.1.1": 1,
  "mpromise@0.5.4": 1,
  "mquery@1.6.3": 1,
  "ms@0.6.2": 1,
  "ms@0.7.1": 9,
  "ms@0.7.3": 1,
  "ms@2.0.0": 2,
  "muri@1.0.0": 1,
  "nan@2.10.0": 1,
  "negotiator@0.2.8": 1,
  "negotiator@0.4.9": 1,
  "negotiator@0.5.3": 1,
  "node-uuid@1.4.8": 1,
  "nopt@2.2.1": 1,
  "nopt@3.0.6": 1,
  "normalize-package-data@2.3.5": 2,
  "normalize-path@2.0.1": 2,
  "npmconf@0.0.24": 1,
  "number-is-nan@1.0.0": 6,
  "nyc@6.6.1": 1,
  "oauth-sign@0.4.0": 1,
  "oauth-sign@0.8.2": 1,
  "object-assign@4.1.0": 1,
  "object.omit@2.0.0": 2,
  "on-finished@2.1.0": 1,
  "on-finished@2.2.1": 4,
  "on-finished@2.3.0": 1,
  "on-headers@1.0.1": 1,
  "once@1.1.1": 1,
  "once@1.3.3": 9,
  "once@1.4.0": 6,
  "only-shallow@1.2.0": 1,
  "opener@1.5.1": 1,
  "optimist@0.6.1": 1,
  "optional@0.1.4": 1,
  "optionator@0.8.1": 1,
  "os-homedir@1.0.1": 1,
  "os-locale@1.4.0": 1,
  "osenv@0.0.3": 1,
  "parse-glob@3.0.4": 2,
  "parse-json@2.2.0": 3,
  "parseurl@1.3.2": 3,
  "path-exists@2.1.0": 6,
  "path-is-absolute@1.0.0": 4,
  "path-is-absolute@1.0.1": 3,
  "path-to-regexp@0.1.3": 1,
  "path-type@1.1.0": 2,
  "pify@2.3.0": 5,
  "pinkie-promise@2.0.1": 18,
  "pinkie@2.0.4": 18,
  "pkg-conf@1.1.3": 1,
  "pkg-dir@1.0.0": 1,
  "pkg-up@1.0.0": 1,
  "ports@1.1.0": 1,
  "prelude-ls@1.1.2": 4,
  "preserve@0.2.0": 2,
  "process-nextick-args@2.0.0": 3,
  "proto-list@1.2.4": 1,
  "proxy-addr@1.0.10": 1,
  "pseudomap@1.0.2": 3,
  "psl@1.1.29": 1,
  "punycode@1.4.1": 3,
  "qs@1.2.2": 1,
  "qs@2.2.4": 1,
  "qs@2.4.2": 1,
  "qs@6.3.2": 1,
  "randomatic@1.1.5": 2,
  "range-parser@1.0.3": 3,
  "raw-body@1.3.0": 1,
  "read-pkg-up@1.0.1": 2,
  "read-pkg@1.1.0": 2,
  "readable-stream@1.0.31": 1,
  "readable-stream@1.0.34": 1,
  "readable-stream@1.1.14": 3,
  "readable-stream@2.3.6": 3,
  "regex-cache@0.4.3": 2,
  "regexp-clone@0.0.1": 2,
  "repeat-element@1.1.2": 4,
  "repeat-string@1.5.4": 4,
  "request@2.42.0": 1,
  "request@2.79.0": 1,
  "require-main-filename@1.0.1": 2,
  "resolve-from@2.0.0": 1,
  "resolve@1.1.7": 1,
  "resumer@0.0.0": 1,
  "right-align@0.1.3": 1,
  "rimraf@2.5.2": 2,
  "rimraf@2.6.2": 1,
  "safe-buffer@5.1.2": 7,
  "safer-buffer@2.1.2": 3,
  "semver@1.1.4": 1,
  "semver@5.1.0": 2,
  "send@0.12.3": 2,
  "serve-static@1.9.3": 1,
  "set-blocking@1.0.0": 1,
  "signal-exit@2.1.2": 4,
  "signal-exit@3.0.0": 1,
  "signal-exit@3.0.2": 1,
  "sliced@0.0.5": 2,
  "slide@1.1.6": 1,
  "sntp@0.2.4": 1,
  "sntp@1.0.9": 1,
  "source-map@0.2.0": 1,
  "source-map@0.4.4": 1,
  "source-map@0.5.6": 2,
  "spawn-wrap@1.2.3": 1,
  "spdx-correct@1.0.2": 2,
  "spdx-exceptions@1.0.4": 2,
  "spdx-expression-parse@1.0.2": 2,
  "spdx-license-ids@1.2.1": 4,
  "split@0.2.10": 1,
  "sprintf-js@1.0.3": 7,
  "sshpk@1.15.2": 1,
  "st@0.2.4": 1,
  "stack-utils@0.4.0": 1,
  "stream-buffers@3.0.2": 1,
  "stream-combiner@0.0.4": 1,
  "streamifier@0.1.1": 1,
  "streamsearch@0.1.2": 1,
  "string-width@1.0.1": 3,
  "string_decoder@0.10.31": 5,
  "string_decoder@1.1.1": 3,
  "stringstream@0.0.6": 2,
  "strip-ansi@3.0.1": 6,
  "strip-bom@2.0.0": 5,
  "supports-color@1.3.1": 1,
  "supports-color@2.0.0": 1,
  "supports-color@3.1.2": 1,
  "symbol@0.2.3": 1,
  "tap-mocha-reporter@0.0.27": 1,
  "tap-parser@1.3.2": 2,
  "tap@5.8.0": 1,
  "tape@2.3.0": 1,
  "test-exclude@1.1.0": 1,
  "through@2.3.8": 3,
  "tmatch@2.0.1": 1,
  "tough-cookie@2.3.4": 1,
  "tough-cookie@2.4.3": 1,
  "tunnel-agent@0.4.3": 2,
  "tweetnacl@0.14.5": 2,
  "type-check@0.3.2": 2,
  "type-is@1.5.7": 1,
  "type-is@1.6.16": 1,
  "uglify-js@2.6.2": 1,
  "uglify-to-browserify@1.0.2": 1,
  "underscore@1.8.3": 1,
  "unicode-length@1.0.3": 1,
  "urlgrey@0.4.0": 1,
  "util-deprecate@1.0.2": 3,
  "utils-merge@1.0.0": 2,
  "uuid@3.3.2": 1,
  "validate-npm-package-license@3.0.1": 2,
  "vary@1.0.1": 1,
  "vary@1.1.2": 1,
  "verror@1.10.0": 1,
  "which@1.2.10": 6,
  "which@1.3.1": 1,
  "window-size@0.1.0": 1,
  "window-size@0.2.0": 1,
  "wordwrap@0.0.2": 1,
  "wordwrap@0.0.3": 1,
  "wordwrap@1.0.0": 2,
  "wrap-ansi@2.0.0": 1,
  "wrappy@1.0.2": 22,
  "write-file-atomic@1.1.4": 1,
  "xtend@4.0.1": 1,
  "y18n@3.2.1": 1,
  "yallist@2.0.0": 2,
  "yallist@2.1.2": 1,
  "yargs-parser@2.4.0": 1,
  "yargs@3.10.0": 1,
  "yargs@4.7.1": 1,
}
`;
